Add Vision Transformer + ViTFeatureExtractor

NielsRogge opened the pull request:
# What does this PR do?

This PR includes 2 things:

* it adds the [Vision Transformer (ViT)](https://arxiv.org/abs/2010.11929) by Google Brain. ViT is a Transformer encoder trained on ImageNet. It is capable of classifying images, by placing a linear classification head on top of the final hidden state of the [CLS] token. I converted the weights from the [timm](https://github.com/rwightman/pytorch-image-models) repository, which already took care of converting the weights of the original implementation (which is written in JAX) into PyTorch. Once this model is added, we can also add [DeIT](https://ai.facebook.com/blog/data-efficient-image-transformers-a-promising-new-technique-for-image-classification/) (Data-efficient Image Transformers) by Facebook AI, which improve upon ViT. 

* it provides a design for the `ViTFeatureExtractor` class, which can be used to prepare images for the model. It inherits from `FeatureExtractionMixin` and defines a `__call__` method. It currently accepts 3 types of inputs: PIL images, Numpy arrays and PyTorch tensors. It defines 2 transformations using `torchvision`: resizing + normalization. It then returns a `BatchFeature` object with 1 key, namely `pixel_values`. 

Demo notebook of combination of `ViTForImageClassification` + `ViTFeatureExtractor`: https://colab.research.google.com/drive/16TCM-tJ1Mfhs00Qas063kWZmAtVJcOeP?usp=sharing

Compared to NLP models (which accept `input_ids`, `attention_mask` and `token_type_ids`), this model only accepts `pixel_values`. The model itself then converts these pixel values into patches (in case of ViT) in the `ViTEmbeddings` class. 

## Help needed

Would be great if you can help me with the following tasks:

- [x] Add and improve tests. Currently I have defined the following tests: `test_modeling_vit.py`, `test_feature_extraction_vit.py`. However, for the former, since ViT does not use `input_ids`/`input_embeds`, some tests are failing, so I wonder whether it should use all tests defined in `test_modeling_common.py`. For the latter, I also need some help in creating random inputs to test the feature extractor on. 
- [x] Add support for `head_mask` in the forward of `ViTModel`. Possibly remove `attention_mask`?
- [x] Run `make fix-copies` (doesn't work right now for me on Windows)
- [x] Remove the `is_decoder` logic from `modeling_vit.py` (since the model was created using the CookieCutter template). I assume that things such as `past_key_values` are not required for an encoder-only model. 

## Who can review?

@patrickvonplaten @LysandreJik @sgugger

NielsRogge commented on src/transformers/models/vit/feature_extraction_vit.py:
```diff
@@ -0,0 +1,142 @@
+                tensor.
+        """
+        # Input type checking for clearer error
+        assert (
+            isinstance(images, PIL.Image.Image)
+            or isinstance(images, np.ndarray)
+            or isinstance(images, torch.Tensor)
+            or (
+                (
+                    isinstance(images, (list, tuple))
+                    and (
+                        len(images) == 0
+                        or (
+                            isinstance(images[0], PIL.Image.Image)
+                            or isinstance(images[0], np.ndarray)
+                            or isinstance(images[0], torch.Tensor)
+                        )
+                    )
+                )
+            )
+        ), (
+            "Images must of type `PIL.Image.Image`, `np.ndarray` or `torch.Tensor` (single example),"
+            "`List[PIL.Image.Image]`, `List[np.ndarray]` or `List[torch.Tensor]` (batch of examples)."
+        )
```
@sgugger pointed out that this should be updated to be more easy to read.

patil-suraj commented:
Hey @NielsRogge 


> Add and improve tests. Currently I have defined the following tests: test_modeling_vit.py, test_feature_extraction_vit.py. However, for the former, since ViT does not use input_ids/input_embeds, some tests are failing, so I wonder whether it should use all tests defined in test_modeling_common.py. For the latter, I also need some help in creating random inputs to test the feature extractor on.


Some common modeling test depend on the specific argument names, (`input_ids`, `input_embeds`). You could just override such tests in your test class and use the correct argument names. For example  the `test_forward_signature` test
expects `inputs_ids`, so it should be overridden in your class to expect `input_values`.  

Also, the tests for `input_embeds` (for example `test_inputs_embeds`) can be skipped since `ViT` does not use those.  Agin just overrides the test and use `pass` in the method body.

You could use the modeling tests of `Wav2Vec2` and `Speech2Text` for reference since those models also have different arguments.

patil-suraj commented:
I like the overall design `ViTFeatureExtractor`. Regrading the import `ViTFeatureExtractor`
I think it should be always imported in the __init__ files, and instead, `ViTFeatureExtractor`  could check for `torchvision` and raise if it’s not installed. Otherwise, the TF tests on CI will fail because they won’t be able to import `ViTFeatureExtractor`  as we don’t install `torchvision` in TF tests.

We should also add the `torchvision`  dependency in the setup.py file as `extras["vision"]` and also add it in `config.yaml` for CI

LysandreJik commented on src/transformers/models/vit/modeling_vit.py:
```diff
@@ -0,0 +1,642 @@
+        super().__init__()
+        image_size = to_2tuple(image_size)
+        patch_size = to_2tuple(patch_size)
+        num_patches = (image_size[1] // patch_size[1]) * (image_size[0] // patch_size[0])
```
A comment here would be helpful

sgugger commented on src/transformers/__init__.py:
```diff
@@ -1410,6 +1426,7 @@
         TransfoXLCorpus,
         TransfoXLTokenizer,
     )
+    from .models.vit import VIT_PRETRAINED_CONFIG_ARCHIVE_MAP, ViTConfig, ViTFeatureExtractor
```
```suggestion
    from .models.vit import VIT_PRETRAINED_CONFIG_ARCHIVE_MAP, ViTConfig
```

sgugger commented on src/transformers/models/vit/configuration_vit.py:
```diff
@@ -0,0 +1,141 @@
+# coding=utf-8
+# Copyright Google AI and The HuggingFace Inc. team. All rights reserved.
```
```suggestion
# Copyright 2021 Google AI and The HuggingFace Inc. team. All rights reserved.
```

sgugger commented on src/transformers/models/vit/convert_vit_timm_to_pytorch.py:
```diff
@@ -0,0 +1,230 @@
+
+def remove_classification_head_(state_dict):
+    ignore_keys = [
+        "norm.weight",
+        "norm.bias",
+        "head.weight",
+        "head.bias",
```
This can all fit in one line.

NielsRogge replied:
I know, it's `make style` that does this.

sgugger replied:
Just tested locally and it did not change the line
```
    ignore_keys = ["norm.weight", "norm.bias", "head.weight", "head.bias"]
```

sgugger commented on src/transformers/models/vit/feature_extraction_vit.py:
```diff
@@ -0,0 +1,136 @@
+
+        Args:
+            images (:obj:`PIL.Image.Image`, :obj:`np.ndarray`, :obj:`torch.Tensor`, :obj:`List[PIL.Image.Image]`, :obj:`List[np.ndarray]`, :obj:`List[torch.Tensor]`):
+                The image or batch of images to be prepared. Each image can be a PIL image, numpy array or a Torch
+                tensor. In case of a numpy array/Torch tensor, each image should be of shape (C, H, W), where C is a
```
```suggestion
                The image or batch of images to be prepared. Each image can be a PIL image, numpy array or a PyTorch
                tensor. In case of a NumPy array/PyTorch tensor, each image should be of shape (C, H, W), where C is a
```

sgugger commented on src/transformers/models/vit/feature_extraction_vit.py:
```diff
@@ -0,0 +1,136 @@
+import numpy as np
+import torch
+from PIL import Image
+from torchvision import transforms as T
```
I'm not super fan of the one-letter import and we try to avoid it when possible. Here we could just import the things we need from `torchvision.transforms`: `ToPILImage`, `Resize`, `ToTensor`, `Normalize` and `Compose`.

sgugger commented on src/transformers/models/vit/modeling_vit.py:
```diff
@@ -0,0 +1,642 @@
+]
+
+
+# Copied from
+# https://github.com/rwightman/pytorch-image-models/blob/b9bd960a032c75ca6b808ddeed76bee5f3ed4972/timm/models/layers/helpers.py
+# From PyTorch internals
```
Let's keep `# Copied from` for our internal copies or my script is going to explode ;-)

```suggestion
# Inspired by
# https://github.com/rwightman/pytorch-image-models/blob/b9bd960a032c75ca6b808ddeed76bee5f3ed4972/timm/models/layers/helpers.py
```

sgugger commented on src/transformers/models/vit/modeling_vit.py:
```diff
@@ -0,0 +1,642 @@
+
+class ViTEmbeddings(nn.Module):
+    """
+    Construct the cls token, position and patch embeddings.
```
```suggestion
    Construct the CLS token, position and patch embeddings.
```

sgugger commented on src/transformers/models/vit/modeling_vit.py:
```diff
@@ -0,0 +1,642 @@
+    """
+    Construct the cls token, position and patch embeddings.
+
+    Based on timm implementation, which can be found here:
+    https://github.com/rwightman/pytorch-image-models/blob/master/timm/models/vision_transformer.py
```
Nit: put this in comment, not the docstring.

LysandreJik commented on src/transformers/models/vit/feature_extraction_vit.py:
```diff
@@ -0,0 +1,136 @@
+        Main method to prepare for the model one or several image(s).
+
+        Args:
+            images (:obj:`PIL.Image.Image`, :obj:`np.ndarray`, :obj:`torch.Tensor`, :obj:`List[PIL.Image.Image]`, :obj:`List[np.ndarray]`, :obj:`List[torch.Tensor]`):
```
It seems important to note that objects which are not `PIL.Image.Image` will be converted to PIL images, before being converted back to torch tensors.

LysandreJik replied:
Also, this feature extractor returns torch tensors. I think one of the expected behaviors of our pre-processors (tokenizers or feature extractors) are to be framework-agnostic, even if no other framework is supported.

The idea is that even if users don't use our implementation of ViT, they can still benefit from the feature extractor.

At the very least it should be explicitely mentioned in the documentation that it will return torch tensors, but I'd much rather it be framework agnostic and have a `return_tensors` flag!

NielsRogge replied:
Yes, currently it's not framework-agnostic, as discussed on Slack. We are adding only ViT in PyTorch for now, and we are using `torchvision` for the image transformations. In the future, one will be able to use the `return_tensors` flag. 

I'll add a `returns` docstring.

sgugger replied:
Maybe add a clear warning to tell the users the processor will convert numpy arrays and torch tensors to PIL images so the most efficient is to pass PIL images?

sgugger commented on src/transformers/models/vit/modeling_vit.py:
```diff
@@ -0,0 +1,642 @@
+    """
+    Image to Patch Embedding.
+
+    Based on timm implementation, which can be found here:
+    https://github.com/rwightman/pytorch-image-models/blob/master/timm/models/vision_transformer.py
```
Same comment as above.

sgugger commented on src/transformers/models/vit/modeling_vit.py:
```diff
@@ -0,0 +1,642 @@
+    def forward(self, pixel_values):
+        batch_size, num_channels, height, width = pixel_values.shape
+        # FIXME look at relaxing size constraints
+        assert (
+            height == self.image_size[0] and width == self.image_size[1]
+        ), f"Input image size ({height}*{width}) doesn't match model ({self.image_size[0]}*{self.image_size[1]})."
```
Since it's on three lines because of the styling, I would prefer:
```suggestion
        if height != self.image_size[0] or width != self.image_size[1]:
            raise ValueError(f"Input image size ({height}*{width}) doesn't match model ({self.image_size[0]}*{self.image_size[1]}).")
```

sgugger commented on src/transformers/models/vit/modeling_vit.py:
```diff
@@ -0,0 +1,642 @@
+        return x.permute(0, 2, 1, 3)
+
+    def forward(
+        self,
+        hidden_states,
+        head_mask=None,
+        output_attentions=False,
```
Should fit on one line.

sgugger commented on src/transformers/models/vit/modeling_vit.py:
```diff
@@ -0,0 +1,642 @@
+class ViTAttention(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+        self.self = ViTSelfAttention(config)
```
Python accepts `self.self` but this is very error-prone and makes the code harder to read. Please change this to `self.attention` 

I realize this means changing the conversion and the weights you have but we can still make that change now and we have that in BERT, but IMO it was a mistake to have a layer with that name so I really really would like to avoid it in new models.

sgugger commented on src/transformers/models/vit/modeling_vit.py:
```diff
@@ -0,0 +1,642 @@
+        # in ViT, layernorm is also applied after self-attention
+        layer_output = self.layernorm_after(hidden_states)
+
+        # feedforward chunking not working for now
```
Add a TODO?

sgugger commented on src/transformers/models/vit/modeling_vit.py:
```diff
@@ -0,0 +1,642 @@
+            all_hidden_states = all_hidden_states + (hidden_states,)
+
+        if not return_dict:
+            return tuple(
+                v
+                for v in [
+                    hidden_states,
+                    all_hidden_states,
+                    all_self_attentions,
+                ]
+                if v is not None
+            )
```
This should all fit in one line. Or if it does not, at least
```
            return tuple(
                v
                for v in [hidden_states, all_hidden_states, all_self_attentions]
                if v is not None
            )
```

sgugger commented on src/transformers/models/vit/modeling_vit.py:
```diff
@@ -0,0 +1,642 @@
+
+    def _init_weights(self, module):
+        """ Initialize the weights """
+        if isinstance(module, (nn.Linear, nn.Embedding, nn.Conv2d)):
+            # Slightly different from the TF version which uses truncated_normal for initialization
+            # cf https://github.com/pytorch/pytorch/pull/5617
+            module.weight.data.normal_(mean=0.0, std=self.config.initializer_range)
+        elif isinstance(module, nn.LayerNorm):
+            module.bias.data.zero_()
+            module.weight.data.fill_(1.0)
+        if isinstance(module, (nn.Linear, nn.Conv2d)) and module.bias is not None:
+            module.bias.data.zero_()
```
Please see the latest `_init_weights` in BERT and adapt here: the code won't work for PyTorch 1.8 otherwise.

NielsRogge replied:
Thanks!

sgugger commented on src/transformers/models/vit/modeling_vit.py:
```diff
@@ -0,0 +1,642 @@
+
+
+VIT_START_DOCSTRING = r"""
+    This model is a PyTorch `torch.nn.Module <https://pytorch.org/docs/stable/nn.html#torch.nn.Module>`_ sub-class. Use
```
```suggestion
    This model is a PyTorch `torch.nn.Module <https://pytorch.org/docs/stable/nn.html#torch.nn.Module>`_ subclass. Use
```

sgugger reviewed (approved):
Thanks a lot for adding this model! The main problem I have is with the `self.self` for the self-attention. It's there in BERT and there is nothing we can do about it now, but we can still make sure to use another name in newer models!

LysandreJik reviewed (commented):
This looks really good, fantastic job @NielsRogge!

Related to the modelcard:
- The model doesn't have a model card as of now, it would be amazing to have it
- The model configuration on your `nielsr/vit-base-patch16-224` repo has all the labels as "LABELS_{i}", it would be great to have the actual label names! 

Other than that it looks in very good shape! I'm wondering about the feature processor as I understand it, it's not framework agnostic. Also, the AutoModel is a very low hanging fruit when we already have the mapping.

LysandreJik commented on src/transformers/testing_utils.py:
```diff
@@ -240,6 +241,19 @@ def require_torchaudio(test_case):
+    These tests are skipped when torchvision isn't installed.
+
+    """
+    if not is_torchvision_available:
```
Seen with @sgugger, this is missing `()`:
```suggestion
    if not is_torchvision_available():
```

NielsRogge commented on src/transformers/models/auto/modeling_auto.py:
```diff
@@ -470,6 +473,13 @@
     ]
 )
 
+MODEL_FOR_IMAGE_CLASSIFICATION_MAPPING = OrderedDict(
```
I've added it, let me know if it looks ok

NielsRogge commented:
Thanks for the reviews, addressed most of the comments. To do:

- [ ] rename `self.self` to `self.attention` and update conversion script accordingly
- [ ] convert more models, place them under the google namespace
- [ ] add model cards

sgugger commented on docs/source/model_doc/vit.rst:
```diff
@@ -0,0 +1,89 @@
+Vision Transformer (ViT)
```
Why did you remove the copyright? Please put i back.

NielsRogge replied:
Oh sorry, I thought it was a mistake. I didn't see this in tapas.rst, so deleted it.

sgugger commented on docs/source/model_doc/vit.rst:
```diff
@@ -0,0 +1,89 @@
+.. note::
+
+    This is a recently introduced model so the API hasn't been tested extensively. There may be some bugs or slight
+    breaking changes to fix them in the future. Note that the API of :obj:`ViTFeatureExtractor` will change in the
+    future, when more frameworks than just PyTorch will be supported.
```
```suggestion
    breaking changes to fix it in the future. Note that the API of :obj:`ViTFeatureExtractor` might change in the
    future, when more frameworks than just PyTorch are supported.
```

sgugger commented on docs/source/index.rst:
```diff
@@ -319,6 +323,8 @@ TensorFlow and/or Flax.
 +-----------------------------+----------------+----------------+-----------------+--------------------+--------------+
 |       Transformer-XL        |       ✅       |       ❌       |       ✅        |         ✅         |      ❌      |
 +-----------------------------+----------------+----------------+-----------------+--------------------+--------------+
+|             ViT             |       ❌       |       ❌       |       ✅        |         ❌         |      ❌      |
```
This is for a further PR though ;-) But yes, definitely worth a look!

sgugger commented on docs/source/model_doc/vit.rst:
```diff
@@ -0,0 +1,101 @@
+.. note::
+
+    This is a recently introduced model so the API hasn't been tested extensively. There may be some bugs or slight
+    breaking changes to fix it in the future. Note that the API of :obj:`ViTFeatureExtractor` will change in the
```
```suggestion
    breaking changes to fix it in the future. Note that the API of :class:`~transformers.ViTFeatureExtractor` will change in the
```

sgugger commented on src/transformers/__init__.py:
```diff
@@ -1530,6 +1546,9 @@
     else:
         from .utils.dummy_tokenizers_objects import *
 
+    if is_torchvision_available():
+        from .models.vit import ViTFeatureExtractor
+
```
Nit: please put it at the end to have the same order as for the dict before (will make life easier for automated scripts).

sgugger commented on src/transformers/models/vit/convert_vit_timm_to_pytorch.py:
```diff
@@ -0,0 +1,1271 @@
+logger = logging.get_logger(__name__)
+
+# ImageNet 2012 id's to class names
+id2class = {
```
Can we store this in a file in transformers.utils instead (so you would import it from there)? This might be needed for other models in the future.

NielsRogge replied:
I've stored it in a file in transformers.utils named `imagenet_classes.py`. Let me know if this looks ok to you.

sgugger commented on src/transformers/models/vit/convert_vit_timm_to_pytorch.py:
```diff
@@ -0,0 +1,1271 @@
+from pathlib import Path
+
+import torch
+import torchvision.transforms as T
```
As in the feature extraction class, can we avoid the one letter import and jsut import the classes needed?

sgugger commented on src/transformers/models/vit/feature_extraction_vit.py:
```diff
@@ -0,0 +1,140 @@
+            Whether or not to normalize the input with mean and standard deviation.
+        do_resize (:obj:`bool`, `optional`, defaults to :obj:`True`):
+            Whether to resize the input to a certain :obj:`size`.
+        size (:obj:`int`, `optional`, defaults to :obj:`224`):
```
```suggestion
        size (:obj:`int`, `optional`, defaults to 224):
```
No obj for ints or floats.

sgugger commented on src/transformers/models/vit/feature_extraction_vit.py:
```diff
@@ -0,0 +1,140 @@
+    methods. Users should refer to this superclass for more information regarding those methods.
+
+    Args:
+        image_mean (:obj:`int`, defaults to [0.5, 0.5, 0.5]):
```
```suggestion
        image_mean (:obj:`int`, defaults to :obj:`[0.5, 0.5, 0.5]`):
```

sgugger commented on src/transformers/models/vit/feature_extraction_vit.py:
```diff
@@ -0,0 +1,140 @@
+    Args:
+        image_mean (:obj:`int`, defaults to [0.5, 0.5, 0.5]):
+            The sequence of means for each channel, to be used when normalizing images.
+        image_std (:obj:`int`, defaults to [0.5, 0.5, 0.5]):
```
```suggestion
        image_std (:obj:`int`, defaults to :obj:`[0.5, 0.5, 0.5]`):
```

sgugger commented on src/transformers/models/vit/feature_extraction_vit.py:
```diff
@@ -0,0 +1,140 @@
+
+    def __init__(
+        self,
+        image_mean=[0.5, 0.5, 0.5],
+        image_std=[0.5, 0.5, 0.5],
```
Setting defaults to mutable objects in Python is pretty dangerous (see [here](https://docs.python-guide.org/writing/gotchas/) for instance). I think it would be cleaner to have `None` here and then setting the value to `[0.5, 0.5, 0.5]` when storing the attribute at line 64-65 (the doc need not change since it will still be the default value, just set later down the road).

sgugger commented on src/transformers/models/vit/modeling_vit.py:
```diff
@@ -0,0 +1,651 @@
+
+# Based on timm implementation, which can be found here:
+# https://github.com/rwightman/pytorch-image-models/blob/master/timm/models/vision_transformer.py
+
+
```
```suggestion
```

sgugger commented on src/transformers/models/vit/modeling_vit.py:
```diff
@@ -0,0 +1,651 @@
+class ViTEmbeddings(nn.Module):
+    """
+    Construct the CLS token, position and patch embeddings.
+
```
```suggestion
```

sgugger commented on src/transformers/models/vit/modeling_vit.py:
```diff
@@ -0,0 +1,651 @@
+class PatchEmbeddings(nn.Module):
+    """
+    Image to Patch Embedding.
+
```
```suggestion
```

sgugger commented on src/transformers/models/vit/modeling_vit.py:
```diff
@@ -0,0 +1,651 @@
+        head_mask=None,
+        output_attentions=False,
+    ):
+        self_outputs = self.attention(
+            hidden_states,
+            head_mask,
+            output_attentions,
+        )
```
```suggestion
        self_outputs = self.attention(hidden_states, head_mask, output_attentions)
```

sgugger commented on src/transformers/models/vit/modeling_vit.py:
```diff
@@ -0,0 +1,651 @@
+            self.intermediate_act_fn = config.hidden_act
+
+    def forward(self, hidden_states):
+
```
```suggestion
```

sgugger commented on src/transformers/models/vit/modeling_vit.py:
```diff
@@ -0,0 +1,651 @@
+        self.layernorm_before = nn.LayerNorm(config.hidden_size, eps=config.layer_norm_eps)
+        self.layernorm_after = nn.LayerNorm(config.hidden_size, eps=config.layer_norm_eps)
+
+    def forward(
+        self,
+        hidden_states,
+        head_mask=None,
+        output_attentions=False,
+    ):
```
```suggestion
    def forward(self, hidden_states, head_mask=None, output_attentions=False):
```

sgugger commented on src/transformers/models/vit/modeling_vit.py:
```diff
@@ -0,0 +1,651 @@
+                    layer_head_mask,
+                )
+            else:
+                layer_outputs = layer_module(
+                    hidden_states,
+                    layer_head_mask,
+                    output_attentions,
+                )
```
```suggestion
                layer_outputs = layer_module(hidden_states, layer_head_mask, output_attentions)
```

sgugger commented on src/transformers/models/vit/modeling_vit.py:
```diff
@@ -0,0 +1,651 @@
+            Pixel values. Padding will be ignored by default should you provide it. Pixel values can be obtained using
+            :class:`~transformers.ViTFeatureExtractor`. See :meth:`transformers.ViTFeatureExtractor.__call__` for
+            details.
+
```
```suggestion
```

sgugger commented on src/transformers/models/vit/modeling_vit.py:
```diff
@@ -0,0 +1,651 @@
+        # and head_mask is converted to shape [num_hidden_layers x batch x num_heads x seq_length x seq_length]
+        head_mask = self.get_head_mask(head_mask, self.config.num_hidden_layers)
+
+        embedding_output = self.embeddings(
+            pixel_values,
+        )
```
```suggestion
        embedding_output = self.embeddings(pixel_values)
```

sgugger commented on src/transformers/testing_utils.py:
```diff
@@ -240,6 +241,19 @@ def require_torchaudio(test_case):
 
 
+def require_torchvision(test_case):
+    """
+    Decorator marking a test that requires torchvision.
+
+    These tests are skipped when torchvision isn't installed.
+
+    """
```
```suggestion
    """
    Decorator marking a test that requires torchvision. These tests are skipped when torchvision isn't installed.
    """
```

sgugger commented on tests/test_modeling_vit.py:
```diff
@@ -0,0 +1,366 @@
+
+    def prepare_config_and_inputs_for_common(self):
+        config_and_inputs = self.prepare_config_and_inputs()
+        (
+            config,
+            pixel_values,
+            labels,
+        ) = config_and_inputs
```
```suggestion
        config, pixel_values, labels = config_and_inputs
```

sgugger commented on tests/test_modeling_vit.py:
```diff
@@ -0,0 +1,366 @@
+
+    @slow
+    def test_inference_image_classification_head(self):
+        # TODO: replace namespace to google
```
```suggestion
```

sgugger commented on tests/test_modeling_vit.py:
```diff
@@ -0,0 +1,366 @@
+class ViTModelIntegrationTest(unittest.TestCase):
+    @cached_property
+    def default_feature_extractor(self):
+        # TODO: add .from_pretrained()
```
Still TODO I believe?

sgugger reviewed (commented):
There are multiple instances of weird styling. In general we use the 119-chars line to its maximum (you can add a ruler in your IDE to see where it is). Sadly `make style` does not put back code split into several lines back in one line if you are using code copied from another part of the lib as a base (where the split might be justified because there were more objects or longer names in the original) so it has to be done by hand.

LysandreJik reviewed (commented):
Cool, this looks great! Looking forward to seeing @sgugger's take on the feature processor.

Played around with the model a bit, it's fun! Great job on the implementation @NielsRogge!

NielsRogge commented on src/transformers/models/vit/__init__.py:
```diff
@@ -0,0 +1,72 @@
+        from .modeling_vit import (
+            VIT_PRETRAINED_MODEL_ARCHIVE_LIST,
+            ViTForImageClassification,
+            ViTLayer,
```
No, indeed. Removed them.

NielsRogge commented:
I've addressed all comments. Most important updates:

* moved the ImageNet id to classes dict to a new file under transformers.utils named `imagenet_classes`.py.
* added a warning to the `__call__` method of `ViTFeatureExtractor` to indicate that NumPy arrays and PyTorch tensors are converted to PIL images when resizing, so it's most efficient to pass in PIL images.  

The remaining comments which are still open have to do with styling. I seem to have some issues with `make style`. The max_length is set to 119, so not sure what's causing this.
